docs/dependencies.md

```diff
@@ -0,0 +1,287 @@
+
+## Python packages
+
+We manage Python packages via the Python-standard `requirements.txt`
+system and virtualenvs, but there’s a number of interesting details
+about how Zulip makes this system work well for us that are worth
+highlighting.  The system is largely managed by the code in
+`scripts/lib/setup_venv.py`
+
+* **Using `pip` to manage dependencies**.  This is standard in the
+  Python ecosystem, and means we only need to record a list of
+  versions in a `requirements.txt` file to declare what we're using.
+  Since we have a few different installation targets, we maintain
+  several `requirements.txt` format files in the `requirements/`
+  directory (e.g. `dev.txt` for development, `prod.txt` for
+  production, `docs.txt` for ReadTheDocs, `common.txt` for the vast
+  major of packages common to prod and development, etc.).  We use
+  `pip install --no-deps` to ensure we only install the packages we
+  explicitly declare as dependencies.
+* **virtualenv with pinned versions**.  For a large application like
+  Zulip, it is important to ensure that we're always using consistent,
+  predictable versions of all of our Python dependencies.  To ensure
+  this, we install our dependencies in a [virtualenv][] that contains
+  only the packages and versions that Zulip needs, and we always pin
+  exact versions of our dependencies in our `requirements.txt` files.
+  We pin exact versions, not minimum versions, so that installing
+  Zulip won't break if a dependency makes a buggy release.  A side
+  effect is that it's easy to debug problems caused by dependency
+  upgrades, since we're always doing those upgrades with an explicit
+  commit updating the `requirements/` directory.
+* **Caching of virtualenvs and packages**.  To make updating the
+  dependencies of a Zulip installation efficient, we maintain a cache
+  of virtualenvs named by the hash of the relevant `requirements.txt`
+  file.  That way, when re-provisioning a development environment or
+  deploying a new production version with the same Python
+  dependencies, no downloading or installation is required: we just
+  use the same virtualenv.  When the only changes are upgraded
+  versions, we'll use [virtualenv-clone][] to clone the most similar
+  existing virtualenv and then just upgrade the packages needed,
+  making small version upgrades extremely efficient.  And finally, we
+  use `pip`'s built-in caching to ensure that a specific version of a
+  specific package is only downloaded once.
+* **Garbage-collecting caches**.  We have a tool,
+  `scripts/lib/clean-venv-cache`, which will clean old cached
+  virtualenvs that are no longer in use.  In production, the algorithm
+  preserves recent virtualenvs as well as those in use by any current
+  production deployment directory under `/home/zulip/deployments/`.
+  This helps ensure that a Zulip installation doesn't leak large
+  amounts of disk over time.
+* **Pinning versions of indirect depedendencies**.  We "pin" or "lock"
```

Comment:
"dependencies"; also on next line

Reply:
And another in a later section; I recommend just searching for /deped/. :-)

Reply:
Fixed these, and also ran a spellchecker :).